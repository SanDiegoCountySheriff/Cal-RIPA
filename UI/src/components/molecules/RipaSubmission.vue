<template>
  <div>
    <v-progress-linear
      v-if="loading && !submission"
      indeterminate
      color="primary"
      class="tw-mt-3"
    ></v-progress-linear>
    <v-container
      class="submissionDetail--container tw-mt-2"
      fluid
      v-if="submission"
    >
      <v-layout v row wrap>
        <v-flex xs12>
          <v-toolbar flat>
            <v-toolbar-title class="tw-uppercase submissionDetail--titleBar"
              >Submission Details</v-toolbar-title
            >
            <v-btn
              @click="handleBackToSubmissions"
              class="backToSubmissionsBtn"
            >
              Back to Submissions
            </v-btn>
          </v-toolbar>
        </v-flex>
      </v-layout>
      <v-layout v-if="!loading" row class="submissionDetail--header" wrap>
        <v-flex xs3>
          <span class="submissionDetail--header--label">Submission ID:</span>
          <span>{{ submission.submission.id }}</span>
        </v-flex>
        <v-flex xs4>
          <span class="submissionDetail--header--label">Date Submitted:</span>
          <span
            >{{
              format(
                new Date(submission.submission.dateSubmitted),
                'yyyy-MM-dd kk:mm',
              )
            }}<br />(Note: It can take up to 48 hours for stop statuses to
            update below.)</span
          >
        </v-flex>
        <v-flex xs2>
          <span class="submissionDetail--header--label"
            >Submission Stops From Date:</span
          >
          <span>{{
            format(
              new Date(submission.submission.minStopDate),
              'yyyy-MM-dd kk:mm',
            )
          }}</span>
        </v-flex>
        <v-flex xs2>
          <span class="submissionDetail--header--label"
            >Submission Stops To Date:</span
          >
          <span>{{
            format(
              new Date(submission.submission.maxStopDate),
              'yyyy-MM-dd kk:mm',
            )
          }}</span>
        </v-flex>
      </v-layout>
      <v-layout v-if="submission" row wrap>
        <v-flex v-if="submission" xs12>
          <div v-if="submission.summary.length" class="submissionSummary">
            <p v-for="(errorCode, index) in submission.summary" :key="index">
              <span class="label">{{ errorCode.code }}</span>
              <v-btn
                :class="{
                  activeFilter: currentErrorCodeFilter === errorCode.code,
                }"
                @click="handleChangeErrorCodeFilter(errorCode.code)"
                text
                >{{ errorCode.count }}</v-btn
              >
            </p>
          </div>
        </v-flex>
        <v-flex xs12>
          <v-data-table
            class="submissionsStopTable"
            :loading="loading"
            :headers="headers"
            :hide-default-footer="true"
            :items="submission.stops"
            :server-items-length="getTotalStops"
            :search="search"
            :sort-by.sync="sortBy"
            :sort-desc.sync="sortDesc"
          >
            <template v-slot:item.actions="{ item }">
              <v-icon
                v-if="statuses.find(s => s.text === item.status).isEditable"
                small
                class="tw-mr-2"
                @click="editItem(item)"
              >
                mdi-pencil
              </v-icon>
            </template>
            <template v-slot:item.error="{ item }">
              <div class="stopError" v-html="item.error"></div>
            </template>
            <template v-slot:item.edited="{ item }">
              {{ item.isEdited ? 'Yes' : 'No' }}
            </template>
            <template v-slot:no-data>
              <div>No Stops Found in This Submission</div>
            </template>
            <template v-slot:footer>
              <div class="paginationWrapper">
                <p>
                  Items {{ calculateItemsFrom }} - {{ calculateItemsTo }} of
                  {{ submission.submission.recordCount }}
                </p>
                <v-pagination
                  v-model="currentPage"
                  :length="getPaginationLength"
                  @next="handleNextPage"
                  @input="handleJumpToPage"
                  :total-visible="20"
                  @previous="handlePreviousPage"
                ></v-pagination>
                <v-combobox
                  outlined
                  dense
                  v-model="itemsPerPage"
                  :items="itemsPerPageOptions"
                  label="Items per page"
                  @input="handleUpdateItemsPerPage"
                  class="itemsPerPageSelector"
                ></v-combobox>
              </div>
            </template>
          </v-data-table>
        </v-flex>
      </v-layout>
    </v-container>
  </div>
</template>

<script>
import { format } from 'date-fns'
import RipaEditStopMixin from '../mixins/RipaEditStopMixin'
import { SUBMISSION_STATUSES } from '../../constants/stop'

export default {
  name: 'ripa-submission',

  data() {
    return {
      headers: [
        { text: 'Stop', value: 'id', sortName: 'id' },
        { text: 'Status', value: 'status', width: 100, sortName: 'Status' },
        { text: 'Edited', value: 'edited', width: 100, sortName: 'IsEdited' },
        {
          text: 'Error (errors from this submission are highlighted)',
          value: 'error',
          sortable: false,
        },
        { text: 'Actions', value: 'actions', sortable: false },
      ],
      format,
      currentSubmissionLoading: false,
      currentPage: 1,
      itemsPerPageOptions: [10, 25, 50, 100, 250, 500, 1000],
      itemsPerPage: 10,
      search: '',
      statuses: SUBMISSION_STATUSES,
      currentOffset: this.currentPage * this.itemsPerPage,
      sortBy: null,
      sortDesc: true,
      currentErrorCodeFilter: null,
    }
  },

  mixins: [RipaEditStopMixin],

  methods: {
    handleBackToSubmissions() {
      this.$router.push('/admin/submissions')
    },
    handleNextPage() {
      // the pagination component updates the current page
      // BEFORE these are called but this math is based on the
      // current value. So need to subtract 1
      this.$emit('submissionDetailPaginate', {
        id: this.submissionId,
        offset: this.itemsPerPage * (this.currentPage - 1),
        limit: this.itemsPerPage,
        filters: this.getFilterStatus,
      })
    },
    handlePreviousPage() {
      this.$emit('submissionDetailPaginate', {
        id: this.submissionId,
        offset: this.itemsPerPage * (this.currentPage - 1),
        limit: this.itemsPerPage,
        filters: this.getFilterStatus,
      })
    },
    handleSort() {
      this.$emit('submissionDetailPaginate', {
        id: this.submissionId,
        offset: this.itemsPerPage * (this.currentPage - 1),
        limit: this.itemsPerPage,
        filters: this.getFilterStatus,
      })
    },
    handleJumpToPage() {
      this.$emit('submissionDetailPaginate', {
        id: this.submissionId,
        offset: this.itemsPerPage * (this.currentPage - 1),
        limit: this.itemsPerPage,
        filters: this.getFilterStatus,
      })
    },
    handleUpdateItemsPerPage(val) {
      this.itemsPerPage = val
      // calculate the page you SHOULD be on with the new items per page
      const newPage = Math.ceil(this.currentPage / this.itemsPerPage)
      this.currentPage = newPage
      this.$emit('redoSubmissionDetailItemsPerPage', {
        id: this.submissionId,
        limit: this.itemsPerPage,
        offset: this.itemsPerPage * (newPage - 1),
        filters: this.getFilterStatus,
      })
    },
    editItem(item) {
      this.handleEditStopByAdmin(item, window.location.pathname)
    },
    getColumnSortName() {
      const columnName = Array.isArray(this.sortBy)
        ? this.sortBy[0]
        : this.sortBy
      const columnToSort = this.headers.filter(headerObj => {
        return headerObj.value === columnName
      })
      if (columnToSort.length) {
        return columnToSort[0].sortName
      } else {
        return null
      }
    },
    handleChangeErrorCodeFilter(whichCode) {
      if (whichCode !== this.currentErrorCodeFilter) {
        this.currentErrorCodeFilter = whichCode
      } else {
        this.currentErrorCodeFilter = null
      }
      this.$emit('submissionDetailPaginate', {
        id: this.submissionId,
        limit: this.itemsPerPage,
        filters: this.getFilterStatus,
      })
    },
  },

  computed: {
    getTotalStops() {
      if (this.submission) {
        return this.submission.submission.recordCount
      } else {
        return 0
      }
    },
    getPaginationLength() {
      if (this.submission) {
        return Math.ceil(
          this.submission.submission.recordCount / this.itemsPerPage,
        )
      } else {
        return 0
      }
    },
    calculateItemsTo() {
      if (this.currentPage === this.getPaginationLength) {
        return this.submission.submission.recordCount
      } else {
        return this.currentPage * this.itemsPerPage
      }
    },
    calculateItemsFrom() {
      if (this.currentPage === 1) {
        return this.currentPage
      } else {
        return (this.currentPage - 1) * this.itemsPerPage + 1
      }
    },
    getFilterStatus() {
      let filters = null
      if (this.getColumnSortName() !== null) {
        let sortOrder = this.sortDesc
        if (Array.isArray(this.sortDesc)) {
          sortOrder = this.sortDesc[0]
        }
        filters = {
          orderBy:
            // if the column sort name is null, default to sorting by the stop date
            this.getColumnSortName() === null
              ? 'dateSubmitted'
              : this.getColumnSortName(),
          order: sortOrder || sortOrder === undefined ? 'Desc' : 'Asc',
        }
      }
      if (this.currentErrorCodeFilter !== null) {
        filters = {
          ...filters,
          errorCode: this.currentErrorCodeFilter,
        }
      }
      return filters
    },
  },

  watch: {
    submissionId(newValue, oldValue) {
      this.currentSubmissionLoading = true
      if (newValue !== oldValue) {
        this.$emit('loadNewSubmission', newValue)
      }
    },
    sortDesc: function (newValue, oldValue) {
      if (newValue === undefined) {
        // this means you're removing the sort on this column
        this.handleSort()
      } else {
        if (newValue !== oldValue) {
          this.handleSort()
        }
      }
    },
  },

  props: {
    submissionId: {
      type: String,
      default: '',
    },
    submission: {
      type: Object,
      default: () => {},
    },
    loading: {
      type: Boolean,
      default: false,
    },
  },
}
</script>

<style lang="scss">
.submissionDetail--container {
  button.backToSubmissionsBtn {
    margin-left: 20px;
    border: 1px solid #666;
  }
}

.submissionDetail--header {
  padding: 16px;
  span {
    display: block;
    margin-right: 8px;
  }

  span.submissionDetail--header--label {
    font-weight: bold;
  }
}

.submissionSummary {
  display: flex;
  padding: 15px;
  border: 1px solid #333;
  align-items: flex-start;
  justify-content: center;
  margin-bottom: 20px;

  > p {
    flex: 1;
    margin-bottom: 0px;
    display: flex;
    flex-direction: column;
    align-items: center;

    .v-btn {
      font-size: 1.1rem;
      font-weight: normal;
      color: #2196f3;
      max-width: 100px;

      &.activeFilter {
        background: #e3e3e3;
        font-weight: bold;
      }
    }

    span {
      display: block;
      text-align: center;
    }

    span.label {
      font-size: 1.2rem;
      font-weight: bold;
    }
    span.count {
      color: #2196f3;
    }
  }
}

.submissionsStopTable {
  .submissionError--wrapper {
    margin: 0;
    max-width: 400px;
  }

  .stopError p {
    margin: 0px;

    &.currentSubmission {
      color: #2196f3;
    }
  }
}
</style>
