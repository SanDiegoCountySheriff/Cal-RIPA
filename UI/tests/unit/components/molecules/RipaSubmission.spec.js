import RipaSubmission from '@/components/molecules/RipaSubmission.vue'
import { shallowMount, mount } from '@vue/test-utils'
import Vuetify from 'vuetify'

describe('Ripa Submission', () => {
  let vuetify
  let wrapper
  let defaultSubmission

  beforeEach(() => {
    vuetify = new Vuetify()
    defaultSubmission = {
      stops: [],
      submission: {
        id: 1,
        dateSubmitted: '2022-03-11T00:00:00',
        errorCount: 0,
        maxStopDate: '2022-02-14T00:00:00',
        minStopDate: '2022-02-14T00:00:00',
        officerId: '1',
        officerName: 'Officer Name',
        recordCount: 1,
      },
      summary: [],
    }
  })

  const factory = propsData => {
    return shallowMount(RipaSubmission, {
      vuetify,
      propsData: {
        ...propsData,
      },
    })
  }

  it('should match snapshot', () => {
    wrapper = mount(RipaSubmission, {
      vuetify,
      propsData: {
        submission: defaultSubmission,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it.todo('should handle back to submission')

  it.todo('should handle next page')

  it.todo('should handle previous page')

  it.todo('should handle sort')

  it.todo('should handle jump to page')

  it.todo('should handle update items per page')

  it.todo('should edit item')

  it.todo('should get column sort name')

  it.todo('should handle change error code filter')

  it.todo('should get total stops')

  it.todo('should get pagination length')

  it.todo('should calculate items to')

  it.todo('should calculate items from')

  it.todo('should get filter status')

  it.todo('should watch submissionId')

  it.todo('should watch sortDesc')
})
