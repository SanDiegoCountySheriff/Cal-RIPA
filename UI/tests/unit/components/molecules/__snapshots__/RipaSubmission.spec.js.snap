// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ripa Submission should match snapshot 1`] = `
<div class="container submissionDetail--container tw-mt-2 container--fluid">
  <!---->
  <div class="layout v row">
    <div class="flex xs12">
      <header class="v-sheet theme--light v-toolbar v-toolbar--flat" style="height: 64px;">
        <div class="v-toolbar__content" style="height: 64px;">
          <div class="v-toolbar__title tw-uppercase submissionDetail--titleBar">Submission Details</div> <button type="button" class="backToSubmissionsBtn v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default"><span class="v-btn__content">
          Back to Submissions
        </span></button>
        </div>
      </header>
    </div>
  </div>
  <div class="layout submissionDetail--header row">
    <div class="flex xs3"><span class="submissionDetail--header--label">Submission ID:</span> <span>1</span></div>
    <div class="flex xs4"><span class="submissionDetail--header--label">Date Submitted:</span> <span>2022-03-11 24:00<br>(Note: It can take up to 48 hours for stop statuses to update
        below.)</span></div>
    <div class="flex xs2"><span class="submissionDetail--header--label">Submission Stops From Date:</span> <span>2022-02-14 24:00</span></div>
    <div class="flex xs2"><span class="submissionDetail--header--label">Submission Stops To Date:</span> <span>2022-02-14 24:00</span></div>
  </div>
  <div class="layout row">
    <div class="flex xs12">
      <!---->
    </div>
    <div class="flex xs12">
      <div class="v-data-table submissionsStopTable v-data-table--has-bottom theme--light">
        <div class="v-data-table__wrapper">
          <table>
            <colgroup>
              <col class="">
              <col class="">
              <col class="">
              <col class="">
              <col class="">
            </colgroup>
            <thead class="v-data-table-header">
              <tr>
                <th role="columnheader" scope="col" aria-label="Stop: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Stop</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
                <th role="columnheader" scope="col" aria-label="Status: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable" style="width: 100px; min-width: 100px;"><span>Status</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
                <th role="columnheader" scope="col" aria-label="Edited: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable" style="width: 100px; min-width: 100px;"><span>Edited</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
                <th role="columnheader" scope="col" aria-label="Error (errors from this submission are highlighted)" class="text-start"><span>Error (errors from this submission are highlighted)</span></th>
                <th role="columnheader" scope="col" aria-label="Actions" class="text-start"><span>Actions</span></th>
              </tr>
            </thead>
            <tbody>
              <tr class="v-data-table__empty-wrapper">
                <td colspan="5">
                  <div>No Stops Found in This Submission</div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <div class="paginationWrapper">
          <p>
            Items 1 - 1 of
            1
          </p>
          <nav role="navigation" aria-label="Pagination Navigation">
            <ul class="v-pagination theme--light">
              <li><button disabled="disabled" type="button" aria-label="Previous page" class="v-pagination__navigation v-pagination__navigation--disabled"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></button></li>
              <li><button type="button" aria-current="true" aria-label="Current Page, Page 1" class="v-pagination__item v-pagination__item--active primary">1</button></li>
              <li><button disabled="disabled" type="button" aria-label="Next page" class="v-pagination__navigation v-pagination__navigation--disabled"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></button></li>
            </ul>
          </nav>
          <div class="v-input itemsPerPageSelector v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-select v-autocomplete">
            <div class="v-input__control">
              <div role="combobox" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-14" class="v-input__slot">
                <fieldset aria-hidden="true">
                  <legend style="width: 0px;"><span class="notranslate">â€‹</span></legend>
                </fieldset>
                <div class="v-select__slot"><label for="input-14" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Items per page</label><input id="input-14" type="text" autocomplete="off">
                  <div class="v-input__append-inner">
                    <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                  </div><input type="hidden" value="10">
                </div>
                <div class="v-menu">
                  <!---->
                </div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
